spec/PeriodicBackgroundSync-index.bs

```diff
@@ -133,43 +146,81 @@ The user agent should cap the duration and frequency of these events to limit re
+    1. If [=periodic sync registration/count of retries=] for |registration| is 0, <dfn>calculate the time to fire of the first attempt of</dfn> the [=periodicsync event=]:
+        1. Let |now|, a timestmap, represent the current time.
+        1. Let |origin| be the [=origin=] associated with |registration|'s [=periodic sync registration/service worker registration=].
+        1. Let |penalty| be a user agent defined penalty to account for the level of user engagement with the origin.
+        1. Let |minimumIntervalForOrigin| be |penalty|*[=minimum periodic sync interval for any origin=].
+        1. Let |delayForOrigin| be the multiple of |minimumIntervalForOrigin| greater than or equal to [=periodic sync registration/minimum interval=].
+        1. Let |timeTillScheduledEventForOrigin|, a number, be the [=periodic sync scheduler/time till the next scheduled periodicsync event=] for |origin|, if any, null otherwise.
+        1. If |timeTillScheduledEventForOrigin| is not null and |timeTillScheduledEventForOrigin| - |delayForOrigin| is less than |minIntervalForOrigin|:
+              1. If |delayForOrigin| is less than or equal to |timeTillScheduledEventForOrigin|, set |delayForOrigin| to |timeTillScheduledEventForOrigin|.
+              1. Else if |delayForOrigin| is less than or equal to |timeTillScheduledEventForOrigin| + |minIntervalForOrigin|, set |delayForOrigin| to |timeTillScheduledEventForOrigin| + |minIntervalForOrigin|.
+        1. Let |timeSinceLastPeriodicSync| be null if [=periodic sync scheduler/time the last periodicsync event  was fired=] is null, else |now| - [=periodic sync scheduler/time the last periodicsync event was fired=].
+        1. If |timeSinceLastPeriodicSync| is null, set |timeToFire| to |delayForOrigin| + |now|.
+        1. Else:
+            1. If |timeSinceLastPeriodicSync| is greater than equal to [=minimum periodic sync interval across origins=], set |timeToFire| to |delayForOrigin| + |now|.
+            1. Else, set |timeTillNextAllowedPeriodicSync| to [=minimum periodic sync interval across origins=] - |timeSinceLastPeriodicSync|. Set |timeToFire| to the maximum of |delayForOrigin| + |now|, and |timeTillNextAllowedPeriodicSync| + |now|.
+        1. Set |registration|'s [=periodic sync registration/time to fire=] to |timeToFire|.
+    1. Else:
+        1. Increment |registration|'s [=periodic sync registration/count of retries=].
+        1. If |registration|'s [=periodic sync registration/count of retries=] is greater than the [=maximum number of retries=] allowed, set [=periodic sync registration/count of retries=] to 0 and follow the steps to [=calculate the time to fire of the first attempt of=] the periodicsync event.
+        1. Else, Set |registration|'s [=periodic sync registration/time to fire=] to |now| + a small back-off that is exponentially proportional to [=periodic sync registration/count of retries=].
```

Comment:
This isn't actually retrying the event, rather deciding on the appropriate back-off time for the retry. The "time to fire" is calculated differently  is different based on the retry-count so far.
I have moved out the incrementing of count of retries to "Fire a periodicsync event".

Reply:
Right now these steps are completely different for retries, which makes them feel like they should be elsewhere. It feels like it would be much simpler if "fire a periodicsync event" handled this.

For instance, retries bypass "minimum periodic sync interval across origins" right? They're scheduled really differently.

Something like this in "fire a periodicsync event" perhaps:

1. Set registration’s state to "firing".
1. Let retryCount be 0.
1. While true:
    1. Let continue be false.
    1. Let success be false.
    1. If registration’s state is "reregistered-while-firing", set retryCount to 0.
    1. Check if the registration is still there, and abort if it isn't.
    1. Fire functional event "periodicsync" … then run the following steps with dispatchedEvent:
        1. Let waitUntilPromise be the result of waiting for all of dispatchedEvent’s extend lifetime promises.
        1. React to fulfillment of waitUntilPromise with the following steps:
            1. Set success to true.
            1. Set continue to true.
        1. React to rejection of waitUntilPromise with the following steps:
            1. Set continue to true.
     1. Wait for continue to be true.
     1. If success is true or retryCount is the maximum number of retries, then:
         1. Set registration’s state to "pending".
         1. Calculate a time to fire for registration.
         1. Abort these steps.
     1. Increment retryCount.
     1. Wait some sort of back-off time based on retryCount.

Reply:
I like this, thanks! Updated, with the inclusion of state == reregistered-while-firing for vii.